igvc_perception/src/projection/main.cpp

```diff
@@ -0,0 +1,342 @@
+#include <cv_bridge/cv_bridge.h>
+#include <image_geometry/pinhole_camera_model.h>
+#include <pcl/kdtree/kdtree_flann.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <pcl_ros/point_cloud.h>
+#include <ros/publisher.h>
+#include <ros/ros.h>
+#include <sensor_msgs/Image.h>
+#include <tf/transform_listener.h>
+#include <igvc_utils/NodeUtils.hpp>
+#include "velodyne_pointcloud/point_types.h"
+
+/**
+ * Reads from velodyne pointcloud and uses a pinhole camera model in order to correlate pixels to 3d points.
+ * Publishes to a pointcloud topic, line_pointcloud, containing all 3d points that are part of a line.
+ */
+
+image_geometry::PinholeCameraModel g_cam_model;
+tf::StampedTransform g_transform_lidar_to_cam;
+tf::StampedTransform g_transform_cam_to_base;
+tf::StampedTransform g_transform_lidar_to_base;
+ros::Publisher g_line_pointcloud_pub;
+ros::Subscriber g_image_sub;
+ros::Subscriber g_lidar_pointcloud_sub;
+sensor_msgs::CameraInfo::ConstPtr g_cam_info;
+cv::Mat g_cv_img;
+int g_resize_width;
+int g_resize_height;
+std::string g_line_topic;
+std::string g_pointcloud_topic;
+std::string g_lidar_topic;
+
+/**
+ * Finds the two nearest neighbors to the input point, constraining the nearest neighbors
+ * to be from different lidar channels.
+ *
+ * @param point The point to find the nearest neighbors of
+ * @param kdtree_list List of 16 kdtrees, one for each lidar channel
+ * @param nearest1 The out parameter for the nearest pixel
+ * @param nearest2 the out parameter for the second nearest pixel from a different channel
+ */
+
+void findNearestNeighbors(pcl::PointXY point, const std::vector<pcl::KdTreeFLANN<pcl::PointXY>> &kdtree_list,
+                          pcl::PointXY &nearest1, pcl::PointXY &nearest2)
+{
+  float nearest_1_dist = FLT_MAX;
+  float nearest_2_dist = FLT_MAX;
```

Comment:
didn't catch this before, but I think using `std::numeric_limits<float>::max()` is more readable here.